src/platforms/common/stm32/serialno.c

```diff
@@ -64,3 +72,32 @@ void read_serial_number(void)
 #endif
 	serial_no[DFU_SERIAL_LENGTH - 1] = '\0';
 }
+
+static char *utoa_upper(unsigned value, char *str, int base)
```

Comment:
Please let's not use raw types and add some `const` in here - base should be unsigned as it cannot be negative, value should probably be `uint32_t`.

Reply:
What type should the function return -- a pointer to const char, i.e. `static const char *utoa_upper`? I remember that functions can't return const entities, and the compiler errors out on me trying to make it `char *const`. Other arguments updated. `value` is callee-clobbered, `base` is not -- but I only adjust the definition for elementary types, not the declaration/prototype (callers couldn't care less and will have to pass the element via regs anyways)

Reply:
Because we're not using the return value, we could make it `void`, but given `str` will have to be `const char *`, we can return `char *` still if we're returning `str + offset;` to allow the result to be directly used to tell how many bytes were written and chain.

Reply:
Applied suggestion, now this function is chainable on consecutive conversions into an auto-advancing buffer pointer.

Reply:
`base` still needs changing from a raw type and signed (it cannot be negative, it's small - a good choice for `uint8_t` or another limited range unsigned type), but otherwise this looks much better.